unittests/AST/AutoDiffIndexSubsetTests.cpp

```diff
@@ -158,3 +158,117 @@ TEST(AutoDiffIndexSubset, Insertion) {
   EXPECT_EQ(indices1->adding(3, ctx.Ctx),
             AutoDiffIndexSubset::get(ctx.Ctx, 5, {0, 2, 3, 4}));
 }
+
+TEST(AutoDiffIndexSubset, Lowering) {
+  TestContext testCtx;
+  auto &C = testCtx.Ctx;
+  // ((T, T)) -> ()
+  EXPECT_EQ(
+      autodiff::getLoweredParameterIndices(
+          AutoDiffIndexSubset::get(C, 1, {0}),
+          FunctionType::get({
+              FunctionType::Param(
+                  TupleType::get({C.TheAnyType, C.TheAnyType}, C))},
+              C.TheEmptyTupleType)),
+      AutoDiffIndexSubset::get(C, 2, {0, 1}));
+  // ((), (T, T)) -> ()
+  EXPECT_EQ(
+      autodiff::getLoweredParameterIndices(
+          AutoDiffIndexSubset::get(C, 2, {1}),
+          FunctionType::get({
+              FunctionType::Param(C.TheEmptyTupleType),
+              FunctionType::Param(
+                  TupleType::get({C.TheAnyType, C.TheAnyType}, C))},
+                                 C.TheEmptyTupleType)),
+      AutoDiffIndexSubset::get(C, 2, {0, 1}));
+  // (T, (T, T)) -> ()
+  EXPECT_EQ(
+    autodiff::getLoweredParameterIndices(
+      AutoDiffIndexSubset::get(C, 2, {1}),
+      FunctionType::get({
+          FunctionType::Param(C.TheAnyType),
+          FunctionType::Param(
+            TupleType::get({C.TheAnyType, C.TheAnyType}, C))},
+        C.TheEmptyTupleType)),
+    AutoDiffIndexSubset::get(C, 3, {1, 2}));
+  // (T, (T, T)) -> ()
+  EXPECT_EQ(
+    autodiff::getLoweredParameterIndices(
+      AutoDiffIndexSubset::get(C, 2, {0, 1}),
+      FunctionType::get({
+          FunctionType::Param(C.TheAnyType),
+          FunctionType::Param(
+            TupleType::get({C.TheAnyType, C.TheAnyType}, C))},
+        C.TheEmptyTupleType)),
+    AutoDiffIndexSubset::get(C, 3, {0, 1, 2}));
+  // (T, ((T, T)), (T, T), T) -> ()
+  EXPECT_EQ(
+    autodiff::getLoweredParameterIndices(
+      AutoDiffIndexSubset::get(C, 4, {0, 1, 3}),
+      FunctionType::get({
+          FunctionType::Param(C.TheAnyType),
+          FunctionType::Param(
+              TupleType::get({
+                  TupleType::get({C.TheAnyType, C.TheAnyType}, C)}, C)),
+          FunctionType::Param(
+            TupleType::get({C.TheAnyType, C.TheAnyType}, C)),
+          FunctionType::Param(C.TheAnyType)},
+        C.TheEmptyTupleType)),
+    AutoDiffIndexSubset::get(C, 6, {0, 1, 2, 5}));
+  // Method (T) -> ((T, T), (T, T), T) -> ()
+  // TODO(TF-874): Fix this unit test.
```

Comment:
Emphasizing commit message from https://github.com/apple/swift/pull/27514/commits/195c5bbad45f6d55dfd0ad373856a375e5936913:

> This test failed with both current logic and WIP logic from
> https://github.com/rxwei/swift/tree/parameter-indices.
> 
> Stdlib compilation fails with the WIP logic.

Reply:
Is this because `autodiff::getLoweredParameterIndices` is not handling currying correctly?

Reply:
I'm not sure. I think it makes sense to revisit this as part of [TF-874](https://bugs.swift.org/projects/TF/issues/TF-874).